checker/src/org/checkerframework/checker/linear/LinearAnnotatedTypeFactory.java

```diff
@@ -1,93 +1,88 @@
+    @Override
+    public CFTransfer createFlowTransferFunction(
+            CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis) {
+        return new LinearFlow(analysis);
+    }
 
-    /**
-     * Performs flow-sensitive analysis to mark reference types {@code Linear} as {@code Unusable}
-     * once they are used up.
-     *
-     * <p>A {code Linear} type is "used up" once the reference is mentioned, as an {@link
-     * IdentifierTree}.
-     */
-    /*
-    private class LinearFlow extends DefaultFlow<DefaultFlowState> {
-        public LinearFlow(BaseTypeChecker checker, CompilationUnitTree root,
-                Set<AnnotationMirror> annotations, AnnotatedTypeFactory factory) {
-            super(checker, root, annotations, factory);
-        }
+    private class LinearFlow extends CFTransfer {
+        private final AnnotatedTypeFactory factory;
 
-        /**
-         * Case 2: add {@code Unusable} to node type, if it is {@code Linear}.
-         *
-        @Override
-        public Void visitIdentifier(IdentifierTree node, Void p) {
-            super.visitIdentifier(node, p);
-            markAsUnusableIfLinear(node);
-            return null;
+        LinearFlow(CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis) {
+            super(analysis);
+            factory = analysis.getTypeFactory();
         }
 
-        /**
-         * If the node is of type {@code Linear}, then transit its type
-         * into an {@code Unusable} type.
-         *
-         * The method should be called on every instance of a tree
-         * that causes the reference to be "used up".
-         *
-        private void markAsUnusableIfLinear(ExpressionTree node) {
-            if (!LinearVisitor.isLocalVarOrParam(node)) {
-                return;
-                }
+        @Override
+        public TransferResult<CFValue, CFStore> visitMethodInvocation(
+                MethodInvocationNode node, TransferInput<CFValue, CFStore> in) {
+            TransferResult<CFValue, CFStore> result = super.visitMethodInvocation(node, in);
 
-            Element elem = TreeUtils.elementFromUse(node);
-            assert elem != null;
-            if (this.flowState.vars.contains(elem)) {
-                int idx = this.flowState.vars.indexOf(elem);
-                if (this.flowState.annos.get(LINEAR, idx)) {
-                    this.flowState.annos.set(UNUSABLE, idx);
-                    this.flowState.annos.clear(LINEAR, idx);
+            @Nullable Tree receiverTree = node.getTarget().getReceiver().getTree();
+            if (receiverTree != null) {
+                AnnotatedTypeMirror type = factory.getAnnotatedType(receiverTree);
+                if (type.isAnnotatedInHierarchy(LINEAR)) {
+                    return new RegularTransferResult<>(
+                            analysis.createSingleAnnotationValue(
+                                    UNUSABLE, result.getResultValue().getUnderlyingType()),
```

Comment:
The refined type created must be a subtype of the declared type.  If it is not, the annotation isn't applied.  Because `@Unsable` is top, this line does change any types.

Reply:
Ok, so all I need to do is remove the `@Linear` annotation from the type, I guess?!

Reply:
I don't get it... the intention here is to make the receiver get the `@Unusable` annotation. That's the top of the hierarchy, but obviously not the default (the default is `@Normal`). So, I don't want to refine, but to widen the type, so as to make the type of the receiver be annotated with `@Unusable`. The concept of sub-typing here seems a little backwards, but I believe my reasoning is correct, isn't it?

Reply:
Yes, you are trying to widen the type here.  But if you widen the type farther than the declared type, the widened type is ignored.

```
   @Linear String s = "linear";
   String t = s.toUpperCase();
```
After `s.toUpperCase()`, the type of `s` should be `@Unusable`, but the declared type of `s` is `@Linear` which is a subtype of `@Unusable`, so the AnnotatedTypeFactory ignores the type produced by dataflow.

Reply:
So it's impossible to widen the type? That makes it impossible to implement this checker then. At least using the flow framework.

Reply:
Right that's what I was trying to say with the main comment on my review https://github.com/typetools/checker-framework/pull/1784#pullrequestreview-91657594.  I think it is possible to design a Linear Checker that can be implemented, but I haven't thought about what the rules/qualifier hierarchy should be.

Reply:
What if I reverse the type hierarchy, so `@Linear` is at the bottom? That would make assigning `@Linear` variables to something else impossible, but that's fine because normally, that's probably what you actually want (that would ensure concurrency safety for the linear variable, similar to how Pony uses capabilities to forbid aliases, which helps make concurrency safe)!
Not sure if I am not seeing some unintended consequence of that, but I will try that if you don't see anything completely wrong with that reasoning :)

Reply:
`@Linear` is bottom.

Reply:
The concept of sub-typing is artificially being imposed on these annotations, so it's tricky to reason about it. Is it possible to just not use the sub-typing relationship at all and just implement the logic using transitions from one type to another? If so, what's the best approach to do that?

Reply:
You could use just the dataflow framework, that's the classes in `org.checkerframework.dataflow`), write some sort of aliasing prevention analysis.  [NullAway](https://github.com/uber/NullAway) and [Error Prone](http://errorprone.info/) both use the dataflow framework for there analyses.  (Dataflow manual is [here](https://checkerframework.org/manual/checker-framework-dataflow-manual.pdf).
  
But if you want to use a checker, then you need a proper type hierarchy.  I think it is possible to design such a hierarchy for the Linear Checker. I just haven't spent the time to come up with the rules.

Reply:
Thanks @smillst I will see if I can find time to read the full paper on the weekend or some other day. This is fun, but I don't have infinite time to spend on it, unfortunately!

BTW reversing the type hierarchy causes that infamous error again and I have no idea why:

```
[type.invalid.annotations.on.use] invalid type: annotations [@Linear] conflict with declaration of type (@com.athaydes.checker.linear.qual.Linear :: java.lang.Object)
```

But nevermind, it seems clear this approach is not going anywhere.